src/hotspot/share/opto/predicates.hpp

```diff
@@ -263,6 +265,42 @@ class RuntimePredicate : public StackObj {
   static bool is_success_proj(Node* node, Deoptimization::DeoptReason deopt_reason);
 };
 
+// A Template Assertion Predicate Bool represents the BoolNode for the initial value or the last value of a
+// Template Assertion Predicate and all the nodes up to and including the OpaqueLoop* nodes.
+class TemplateAssertionPredicateBool : public StackObj {
```

Comment:
The current name targets the "bool" node, and it is not clear that you actually mean to represent the whole assertion-predicate expression. Hence:
`TemplateAssertionPredicateBool` -> `TemplateAssertionPredicateExpression`

Then `could_be_part` makes more sense. Maybe still, you could rename it to `maybe_contains`?
But is it really only "maybe" or "could"? Maybe add some explanation in what cases it looks like it could be part of it, but is actually not?

Also, expression would help to make more sense of clone, since it clones the whole expression!

Reply:
And why not also include the `Opaque4` node below the bool node?